pkg/operator/cockroachdb/controller.go

```diff
@@ -0,0 +1,573 @@
+	Group:   cockroachdbv1alpha1.CustomResourceGroup,
+	Version: cockroachdbv1alpha1.Version,
+	Scope:   apiextensionsv1beta1.NamespaceScoped,
+	Kind:    reflect.TypeOf(cockroachdbv1alpha1.Cluster{}).Name(),
+}
+
+type ClusterController struct {
+	context                 *clusterd.Context
+	containerImage          string
+	createInitRetryInterval time.Duration
+}
+
+func NewClusterController(context *clusterd.Context, containerImage string) *ClusterController {
+	return &ClusterController{
+		context:                 context,
+		containerImage:          containerImage,
+		createInitRetryInterval: createInitRetryIntervalDefault,
+	}
+}
+
+type cluster struct {
+	context   *clusterd.Context
+	namespace string
+	spec      cockroachdbv1alpha1.ClusterSpec
+	ownerRef  metav1.OwnerReference
+}
+
+func newCluster(c *cockroachdbv1alpha1.Cluster, context *clusterd.Context) *cluster {
+	return &cluster{
+		context:   context,
+		namespace: c.Namespace,
+		spec:      c.Spec,
+		ownerRef:  clusterOwnerRef(c.Namespace, string(c.UID)),
+	}
+}
+
+func clusterOwnerRef(namespace, clusterID string) metav1.OwnerReference {
+	blockOwner := true
+	return metav1.OwnerReference{
+		APIVersion:         ClusterResource.Version,
+		Kind:               ClusterResource.Kind,
+		Name:               namespace,
+		UID:                types.UID(clusterID),
+		BlockOwnerDeletion: &blockOwner,
+	}
+}
+
+func (c *ClusterController) StartWatch(namespace string, stopCh chan struct{}) error {
+
+	resourceHandlerFuncs := cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.onAdd,
+		UpdateFunc: c.onUpdate,
+		DeleteFunc: c.onDelete,
+	}
+
+	logger.Infof("start watching cockroachdb clusters in all namespaces")
+	watcher := opkit.NewWatcher(ClusterResource, namespace, resourceHandlerFuncs, c.context.RookClientset.CockroachdbV1alpha1().RESTClient())
+	go watcher.Watch(&cockroachdbv1alpha1.Cluster{}, stopCh)
+
+	return nil
+}
+
+func (c *ClusterController) onAdd(obj interface{}) {
+	clusterObj := obj.(*cockroachdbv1alpha1.Cluster)
+	logger.Infof("new cluster %s added to namespace %s", clusterObj.Name, clusterObj.Namespace)
+
+	cluster := newCluster(clusterObj, c.context)
+
+	if err := validateClusterSpec(cluster.spec); err != nil {
+		logger.Errorf("invalid cluster spec: %+v", err)
+		return
+	}
+
+	if err := c.createClientService(cluster); err != nil {
+		logger.Errorf("failed to create client service: %+v", err)
+		return
+	}
+
+	if err := c.createReplicaService(cluster); err != nil {
+		logger.Errorf("failed to create replica service: %+v", err)
+		return
+	}
+
+	if err := c.createPodDisruptionBudget(cluster); err != nil {
+		logger.Errorf("failed to create pod disruption budget: %+v", err)
+		return
+	}
+
+	if err := c.createStatefulSet(cluster); err != nil {
+		logger.Errorf("failed to create stateful set: %+v", err)
+		return
+	}
+
+	// retry to init the cluster until it succeeds or times out
+	err := wait.Poll(c.createInitRetryInterval, createInitTimeout, func() (bool, error) {
```

Comment:
could this be dependent on pulling a new image? If so I wonder if the timeout is long enough.